### ISSUE TITLE
Touchevents missing?

### ISSUE THREAD

**Issue by Falke-Design** [Leaflet/Leaflet#7752]
Are here the touch events missing?
https://github.com/Leaflet/Leaflet/blob/257ca098f1edfe091ec029bcead25f7f9ce2b676/src/layer/vector/Canvas.js#L60-L69

And maybe here?
https://github.com/Leaflet/Leaflet/blob/257ca098f1edfe091ec029bcead25f7f9ce2b676/src/map/Map.js#L1311-L1312

**Comment by mourner** [comment]
No, touch devices still fire mouse events. https://developer.mozilla.org/en-US/docs/Web/API/Touch_events/Supporting_both_TouchEvent_and_MouseEvent

**Comment by Falke-Design** [comment]
Okay thanks, then I was not so wrong with https://github.com/Leaflet/Leaflet/pull/7059#pullrequestreview-796457015 👍

**Comment by Falke-Design** [comment]
@mourner After digging more into this, it looks like mouse and touchevents are **not** fired in the same way. The mouseevents are still there but not usable.
I created a very minimalistic example without any librarys and added touch & mouse events to a DOM element.
The touchevents fired much often than the mouseevents. Also mouseevents are only fired if we click on the element without moveing.

Tested on Firefox Touch Simulation, Firefox Android, Chrome Android

```
<div id="map" style="width: 400px; height: 400px;background: #f00"></div>
<div id="debug" style="position: absolute; top: 0;z-index: 10000000;background: #ffffff;"></div>


var map = document.getElementById('map');

    var events = {}
	function updateHtml(e){
		var obj = events[e.type];
		events[e.type] = obj ? obj + 1 : 1;

		var elm = document.getElementById('debug');

		var text = "";
		for(var type in events) {
			text += type+": "+events[type]+"<br>";
		}
		elm.innerHTML = text;
	}

	map.addEventListener('touchmove',updateHtml)
	map.addEventListener('mousemove',updateHtml)

	map.addEventListener('touchstart',updateHtml)
	map.addEventListener('mousedown',updateHtml)

	map.addEventListener('touchend',updateHtml)
	map.addEventListener('mouseup',updateHtml)
```

Demo: https://florianbischof.net/leaflet/testEvents.html

What do you think?

**Comment by johnd0e** [comment]
> The touchevents fired much often than the mouseevents. Also mouseevents are only fired if we click on the element without moveing.

That is right. Only 'click' produces compatibility mouse events.

**Comment by johnd0e** [comment]
Still I'm not sure that leaflet objects should support touch directly.
You could make some proof-of-concept to make things more obvious.

**Comment by Falke-Design** [comment]
> Still I'm not sure that leaflet objects should support touch directly.

Because of the same reasons as mouseevents are supported. Catching moving.
There are so many projects outside which are using `mousedown`, `mousemove`, `mouseup` listeners (on the leaflet layer). If they want to support toch devices they need to change the code from the leaflet layer listener to the DomEvent listener, because else they can't catch touchevents. But then Canvas will not work (#7835) because the layers don't have own DOM elements anymore and the leaflet layer don't support touchevents.

I will create a PR or update #7835 and then we can look into it. It should not be that much to change.

**Comment by nathanlisch** [comment]
Possibly related to this issue, we have a use case with draggable GeoJSON objects on the map where the drag initiates on mousedown, but on touchscreen mousedown isn't firing until after the touch is lifted. So touching the object and moving around just pans the map, because our drag start event isn't happening.

(we are using leaflet 1.8).

**Comment by Falke-Design** [comment]
@nathanlisch Leaflet 1.8 does not exist ;) 
But yes this looks like that. What do you use fro dragging? `L.Draggable`, custom mouse handlers or a plugin?

**Comment by nathanlisch** [comment]
Sorry, referring to [leaflet 1.8-dev](https://leafletjs.com/download.html), which we're using because https://github.com/Leaflet/Leaflet/pull/7026 fixes another issue we were having.

And yes, we're using L.Draggable.

**Comment by Falke-Design** [comment]
Try the code of this [sample](https://plnkr.co/edit/jBHoKEsNfyCPSq8r). This is working with touchevents. But you need to copy it into a other html file / sandbox because plunkr not support touchevents on theier website.

Does it work for you? Maybe another issue with Safari / IOS?

**Comment by nathanlisch** [comment]
Okay, that does work. My mistake, looks like we are not using L.Draggable. We're using a custom drag handler that we call from a mousedown event on the GeoJSON object. Something like this (sorry if it doesn't make sense, the files are very large, so I'm trying to only paste in the relevant parts):

```
L.Handler.ObjectDrag = L.Handler.extend({
  _onDragStart: function(evt) {
    // ...
  },

  _onDrag: function(evt) {
    // ...
  },

  _onDrag: function(evt) {
    // ...
  },
});

L.Handler.ObjectDrag.makeDraggable = function(layer) {
  layer.dragging = new L.Handler.ObjectDrag(layer);
  return layer;
};

L.Path.prototype.makeDraggable = function() {
  return L.Handler.ObjectDrag.makeDraggable(this);
};
```
And then in the object component (we're using react-leaflet):
```
const ObjectComponent = () => {
  const geoJsonRef = useRef(null);

  const eventHandlers = {
    mousedown: (evt) => {
      // BREAKPOINT
      geoJsonRef.current?.getLayers()[0].dragging.enable();
      geoJsonRef.current?.getLayers()[0].dragging._onDragStart(evt);
    }
  }

  const initDraggableLayer = (feature, layer) => {
    layer.makeDraggable();
  };

  return <GeoJSON
        ref={geoJsonRef}
        eventHandlers={eventHandlers}
        onEachFeature={initDraggableLayer}
    />
}
```

If I put a breakpoint where I commented "BREAKPOINT", on desktop it hits the breakpoint at the beginning of the click as expected, but on touchscreen it hits the breakpoint when ending the touch.

Sorry if this isn't very helpful. If I get time I'll try to reproduce in a sandbox.

**Comment by nathanlisch** [comment]
Btw @Falke-Design, when I switch our leaflet instance to use the branch from your PR above, I'm able to add a touchstart event that triggers at the beginning of the touch, but with the MouseEvent we later on pass the event object into Map.mouseEventToLatLng(), which breaks if I try to pass in the TouchEvent. If this is merged, will there be plans to add similar conversion methods for TouchEvents?

**Comment by Falke-Design** [comment]
Yes your code has two problems:
1. You only listen on mousevents (`mousedown`)  and not on touchevents (`touchstart`)
2. You add the listener on the Leaflet layer, this means touchevents will currently not passed through. This is the exact problem of this issue (/PR)

A workaround for this is to add a DomEvent listener on the DOM-Object instead of adding a listener to the Leaflet layer `L.DomEvent.on(layer._path, 'mousedown touchstart', (e)=>console.log(e))` but this will not work with Canvas.

> Map.mouseEventToLatLng(), which breaks if I try to pass in the TouchEvent

We can think of that while that, you can get the first `touches` and pass this to `mouseEventToLatLng`:
```
var first = e.touches ? e.touches[0] : e;
if (['touchend', 'touchcancel'].indexOf(e.type) > -1) {
	first = e.changedTouches[0];
}
var point = this._map.mouseEventToLayerPoint(first)
```

**Comment by nathanlisch** [comment]
Okay, so using the above PR branch (https://github.com/Leaflet/Leaflet/pull/7842), and replacing the touchevent with its nested changedTouch object, I'm able to get our dragging working on touchscreen. Without the PR changes though, it doesn't recognize touchstart as a valid handler. So this PR definitely helps with our use case.

```
  const eventHandlers = {
    mousedown: (evt) => {
      geoJsonRef.current?.getLayers()[0].dragging.enable();
      geoJsonRef.current?.getLayers()[0].dragging._onDragStart(evt);
    },
    touchstart: (evt) => {
      geoJsonRef.current?.getLayers()[0].dragging.enable();
      geoJsonRef.current?.getLayers()[0].dragging._onDragStart(evt);
    }
  }
```

```
_onDrag: (evt) => {
    if (evt.type === 'touchmove') {
      evt = evt.changedTouches[0];
    }
    // ...
}
```

Thanks for your help @Falke-Design!

**Comment by marcoventer777** [comment]
@Falke-Design any updates? I see you liked what @nathanlisch suggested here: https://github.com/Leaflet/Leaflet/issues/7752#issuecomment-996022628

**Comment by jonkoops** [comment]
I think our plan remains to attempt to refactor Leaflet to entirely rely on the [Pointer Events API](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) instead (hopefully before v2). I do believe that would resolve this issue.

**Comment by adnanerlansyah403** [comment]
> I think our plan remains to attempt to refactor Leaflet to entirely rely on the [Pointer Events API](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) instead (hopefully before v2). I do believe that would resolve this issue.

Hi sir, are there any updates for this? The current version of Leaflet is almost at v2.

**Comment by jonkoops** [comment]
>  The current version of Leaflet is almost at v2.

It is not, we still have to work on and release v2, but I am personally chronically short on spare time to work on it.

**Comment by IvanSanchez** [comment]
> but I am personally chronically short on spare time to work on it.

It's not just you, @jonkoops, not just you.

**Comment by marcopolov2** [comment]
Good approach, just have 2 more questions:
1. When are these listeners destroyed?
2. Will these listeners cross interfere with sibling events?
